crates/video/src/video.rs

```diff
@@ -0,0 +1,1125 @@
+            }
+            QueryControl(control_type) => {
+                let mut response = VideoQueryControlResp::default();
+                response.hdr.type_ = self.cmd_type();
+                response.resp.num = control_type.get_value();
+                response_raw.append(&mut response.to_bytes());
+            }
+            GetParams {
+                queue_type: _,
+                params,
+            } => {
+                let mut response = VideoGetParamsResp::default();
+                response.hdr.type_ = self.cmd_type();
+                response.params = *params;
+                response_raw.append(&mut response.to_bytes());
+            }
+            GetControl(_c) => {
+                let mut response = VideoGetControlResp::default();
+                response.hdr.type_ = self.cmd_type();
+                response_raw.extend_from_slice(response.hdr.as_slice());
+            }
+            ResourceQueue {
+                timestamp,
+                flags,
+                size,
+            } => {
+                let mut response = VideoResourceQueueResp::default();
+                response.hdr.type_ = self.cmd_type();
+                response.timestamp = (*timestamp).into();
+                response.flags = BufferFlags::from_bits_retain(*flags);
+                // Only used for encoder
+                response.size = (*size).into();
+                response_raw.append(&mut response.to_bytes());
+            }
+            OkNoData | SetControl | Error(_) => response_raw.extend_from_slice(
+                VirtioVideoHeader {
+                    type_: self.cmd_type(),
+                    ..Default::default()
+                }
+                .as_slice(),
+            ),
+        };
+        response_raw
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum CmdResponseType {
+    Sync(CmdResponse),
+    AsyncQueue {
+        stream_id: u32,
+        queue_type: QueueType,
+        resource_id: u32,
+    },
+}
+
+/// Virtio Video Formats
+#[derive(Debug, Clone, Copy, PartialEq, Eq, N)]
+pub enum Format {
+    /// Raw formats
+    Argb8888 = 1,
+    Bgra8888,
+    Nv12,
+    Yuv420,
+    Yvu420,
+    /// Coded formats
+    Mpeg2 = 0x1000,
+    Mpeg4,
+    H264,
+    Hevc,
+    Vp8,
+    Vp9,
+    Fwht,
+}
+
+impl Default for Format {
+    fn default() -> Self {
+        Self::Fwht
+    }
+}
+
+/// Virtio Video Controls
+#[derive(Debug, Clone, Copy, PartialEq, N)]
+pub enum ControlType {
+    Bitrate = 1,
+    Profile,
+    Level,
+    ForceKeyframe,
+    BitrateMode,
+    BitratePeak,
+    PrependSpsPpsToIdr,
+}
+
+/// Planes layout
+pub(crate) const VIRTIO_VIDEO_PLANES_LAYOUT_SINGLE_BUFFER: u32 = 1 << 0;
+pub(crate) const VIRTIO_VIDEO_PLANES_LAYOUT_PER_PLANE: u16 = 1 << 1;
+
+/// Queue type
+pub(crate) const VIRTIO_VIDEO_QUEUE_TYPE_INPUT: u32 = 0x100;
+pub(crate) const VIRTIO_VIDEO_QUEUE_TYPE_OUTPUT: u32 = 0x101;
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, N)]
+pub enum QueueType {
+    #[default]
```

Comment:
Interesting, you used it here but not elsewhere :)

Reply:
Because I was learning in the process, and then I didn't revisit everything afterward :)